File: src/ParticleRegion.cpp
/*
  Copyright (C) 2016
      Jakub Krajniak (c) (jkrajniak at gmail.com)
  
  This file is part of ESPResSo++.
  
  ESPResSo++ is free software: you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License, or
  (at your option) any later version.
  
  ESPResSo++ is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.
  
  You should have received a copy of the GNU General Public License
  along with this program.  If not, see <http://www.gnu.org/licenses/>. 
*/

#include "python.hpp"
#include "ParticleRegion.hpp"
#include "storage/Storage.hpp"
#include "iterator/CellListIterator.hpp"

namespace espressopp {

LOG4ESPP_LOGGER(ParticleRegion::theLogger, "ParticleRegion");

ParticleRegion::ParticleRegion(shared_ptr<storage::Storage> _storage, shared_ptr<integrator::MDIntegrator> integrator)
    : storage(_storage), integrator_(integrator), ParticleGroup(), velocity_left_(0.0), velocity_right_(0.0) {
  con_changed = storage->onParticlesChanged.connect
      (boost::bind(&ParticleRegion::onParticlesChanged, this));

  sig_aftIntV1  = integrator_->aftIntV.connect(boost::bind(&ParticleRegion::onParticlesChanged, this));
  sig_aftIntV2  = integrator_->aftIntV.connect(boost::bind(&ParticleRegion::updateRegion, this));
  has_types_ = false;
}

ParticleRegion::~ParticleRegion() {
  con_changed.disconnect();
  sig_aftIntV1.disconnect();
  sig_aftIntV2.disconnect();
}

bool ParticleRegion::has(longint pid) {
  return particles.find(pid) != particles.end();
}

// for debugging purpose
void ParticleRegion::print() {
  std::cout << "####### I have " << active.size() << " active particles";
  std::cout << " region: " << left_bottom_ << " to " << right_top_ << std::endl;
  for(iterator i=begin(); i!=end(); ++i ) {
    std::cout << "pid: " << i->id() << " " << i->type() << std::endl;
  }
}

python::list ParticleRegion::getParticleIDs() {
  python::list particle_ids;
  for(iterator i=begin(); i!=end(); ++i ) {
    particle_ids.append(i->id());
  }

  return particle_ids;
}

void ParticleRegion::onParticlesChanged() {
  LOG4ESPP_DEBUG(theLogger, "ParticleRegion::onParticlesChanges");
  LOG4ESPP_DEBUG(theLogger, "left: " << left_bottom_ << " right: " << right_top_);
  active.clear();

  // Update active list.
  CellList cl = storage->getLocalCells();
  for(espressopp::iterator::CellListIterator cit(cl); !cit.isDone(); ++cit) {
    Particle &p = *cit;
    LOG4ESPP_DEBUG(theLogger, "particle id=" << p.id() << " t= " << p.type() << " p= " << p.position() << " g=" << p.ghost());
    if (!has_types_ || (has_types_ && types_.count(p.type()) == 1)) {
      Real3D pos = p.position();
      bool in_region = true;
      for (int d = 0; d < 3; d++) {
        in_region = in_region && (pos[d] > left_bottom_[d]) && (pos[d] < right_top_[d]);
      }
      if (in_region) {
        LOG4ESPP_DEBUG(theLogger, "insert  p " << p.id() << " in=" << in_region);
        active.insert(std::make_pair(p.id(), &p));
      }
    }
  }
}

void ParticleRegion::updateRegion() {
  real dt = integrator_->getTimeStep();
  Real3D delta_p = dt * velocity_left_;
  left_bottom_ += delta_p;

  delta_p = dt * velocity_right_;
  right_top_ += delta_p;
}

python::tuple ParticleRegion::getRegion() {
  return python::make_tuple(
      python::make_tuple(left_bottom_[0], left_bottom_[1], left_bottom_[2]),
      python::make_tuple(right_top_[0], right_top_[1], right_top_[2])
  );
}

void ParticleRegion::registerPython() {
  using namespace espressopp::python;

  class_<ParticleRegion, shared_ptr<ParticleRegion>, bases<ParticleGroup> >
      ("ParticleRegion", init<shared_ptr<storage::Storage>, shared_ptr<integrator::MDIntegrator> >())
        .def("show", &ParticleRegion::print)
        .def("has", &ParticleRegion::has)
        .def("define_region", &ParticleRegion::defineRegion)
        .def("get_region", &ParticleRegion::getRegion)
        .def("add_type_id", &ParticleRegion::addTypeId)
        .def("remove_type_id", &ParticleRegion::removeTypeId)
        .def("get_particle_ids", &ParticleRegion::getParticleIDs)
        .def("set_v", &ParticleRegion::set_v)
        .def("get_v", &ParticleRegion::get_v)
        .def("size", &ParticleRegion::size);
}


}  // end namespace espressopp

File: src/ParticleRegion.py
#  Copyright (C) 2016
#      Jakub Krajniak (c) (jkrajniak at gmail.com)
#  
#  This file is part of ESPResSo++.
#  
#  ESPResSo++ is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation, either version 3 of the License, or
#  (at your option) any later version.
#  
#  ESPResSo++ is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#  
#  You should have received a copy of the GNU General Public License
#  along with this program.  If not, see <http://www.gnu.org/licenses/>. 


r"""
*****************************
**espressopp.ParticleRegion**
*****************************

Defines a geometrical block and gets the list of particles in that region, updated
dynamicly during the simulation. There is no limit on number of such regions in the
simulation box.
The **ParticleRegion** can store particles of any type or only of the type
defined byt the **add_type_id**.

The particle is in the region whenever :math:`p > l ^ p < r` where
:math:`p` is a position of particle, :math:`l`: is the position of left-bottom-front vertex
of the block and :math:`r` is the position of right-top-rear vertex.

**ParticleRegion** inherits interface from **ParticleGroup**, therefore whenever
**ParticleGroup** is accepted then **ParticleRegion** can also be passed.

The region can be extended or shrinked during the simulation by defining the speed of it. The update
is always made on the right top bottom coordinate of the region and the velocity can be positive or negative. By
default it is 0.0 in all directions.

Example of usage
==================

>>> pr = espressopp.ParticleRegion(storage, espressopp.Real3D(-1, -1, 5), espressopp.Real3D(-1, -1, 11))


.. class:: espressopp.ParticleRegion

      The particle region class.

      .. method:: espressopp.ParticleRegion(storage, left, right)

            Define the particle region.

            :param storage: The storage object.
		    :type storage: espressopp.storage.Storage
		    :param left: The left-bottom-front vertex.
		    :type left: espressopp.Real3D
		    :param right: The right-top-rear vertex.
		    :type right: espressopp.Real3D

	  .. method:: espressopp.ParticleRegion.set_v(vx, vy, vz)

            Set the velocity of which the region will change the dimension. The velocity
            can be positive or negative.

            :param vx: The velocity in x-direction.
            :type vx: float
            :param vy: The velocity in y-direction.
            :type vy: float
            :param vz: The velocity in z-direction.
            :type vz: float

	  .. method:: espressopp.ParticleRegion.get_v()

            Returns the velocity of the changes in region.

            :rtype: tuple

      .. method:: espressopp.ParticleRegion.has(pid)

            Check if given particle of *pid* is in the region.

            :param pid:
            :type pid:
            :rtype: bool

      .. method:: espressopp.ParticleRegion.show()

            Debug purpose, print all particles in the region.


      .. method:: espressopp.ParticleRegion.size()

            Gets the number of particles in the region.

            :rtype: int


      .. method:: espressopp.ParticleRegion.add_type_id(type_id)

            Stores only particles of given type.

            :param type_id: Type id to add
            :type type_id: int


      .. method:: espressopp.ParticleRegion.remove_type_id(type_id)

            Removes type from the list of observed types.

            :param type_id: Type id to remove.
            :type type_id: int


.. function:: espressopp.ParticleRegion(storage, left, right)

		:param storage: The storage object.
		:type storage: espressopp.storage.Storage
		:param left: The left-bottom-front vertex.
		:type left: espressopp.Real3D
		:param right: The right-top-rear vertex.
		:type right: espressopp.Real3D

.. function:: espressopp.ParticleRegion.has(pid)
   Check if given particle of *pid* is in the region.

		:param pid: 
		:type pid: 
		:rtype: 

.. function:: espressopp.ParticleRegion.show()
   Debug purpose, print all particles in the region.
		:rtype: 

.. function:: espressopp.ParticleRegion.size()
   Gets the number of particles in the region.
		:rtype:


.. function:: espressopp.ParticleRegion.add_type_id(type_id)
   Stores only particles of given type.

        :param type_id: Type id to add
        :type type_id: int


.. function:: espressopp.ParticleRegion.remove_type_id(type_id)
   Removes type from the list of observed types.

        :param type_id: Type id to remove.
        :type type_id: int

"""
import _espressopp
from espressopp.esutil import cxxinit

import pmi


class ParticleRegionLocal(_espressopp.ParticleRegion):
    def __init__(self, storage, integrator, left_bottom, right_top):
        if pmi.workerIsActive():
            cxxinit(self, _espressopp.ParticleRegion, storage, integrator)
            self.cxxclass.define_region(self, left_bottom, right_top)

    def add_type_id(self, type_id):
        if pmi.workerIsActive():
            self.cxxclass.add_type_id(self, type_id)

    def remove_type_id(self, type_id):
        if pmi.workerIsActive():
            self.cxxclass.remove_type_id(self, type_id)

    def show(self):
        if pmi.workerIsActive():
            self.cxxclass.show(self)

    def has(self, pid):
        if pmi.workerIsActive():
            return self.cxxclass.has(self, pid)

    def size(self):
        if pmi.workerIsActive():
            return self.cxxclass.size(self)

    def set_v(self, vx, vy, vz, left_right='left'):
        if pmi.workerIsActive():
            self.cxxclass.set_v(self, vx, vy, vz, left_right=='left')

    def get_v(self, left_right='left'):
        if pmi.workerIsActive():
            return self.cxxclass.get_v(self, left_right=='left')

    def get_region(self):
        if pmi.workerIsActive():
            return self.cxxclass.get_region(self)


if pmi.isController:
    class ParticleRegion(object):
        __metaclass__ = pmi.Proxy
        pmiproxydefs = dict(
            cls='espressopp.ParticleRegionLocal',
            pmiinvoke=['get_particle_ids'],
            pmicall=['show', 'has', 'size', 'define_region', 'add_type_id', 'remove_type_id', 'set_v', 'get_v',
                     'get_region']
        )
